Use LoggerMessage.Define in HttpSys

martincostello opened the pull request:
**PR Title**

Use `LoggerMessage.Define()` in HttpSys

**PR Description**

Use `LoggerMessage.Define()` throughout HttpSys instead of `ILogger` extension methods.

Also updates two incorrect event Id field names (but not the event names themselves, as I assume this would be a breaking change).

Addresses #31313

martincostello committed:
Use LoggerMessage.Define in HttpSys

Use LoggerMessage.Define() throughout HttpSys instead of ILogger extension methods.
Also updates two incorrect event Id field names.
Addresses #31313.

martincostello commented on src/Servers/HttpSys/src/NativeInterop/DisconnectListener.Log.cs:
```diff
@@ -0,0 +1,62 @@
+            private static readonly Action<ILogger, Exception?> _disconnectHandlerError =
+                LoggerMessage.Define(LogLevel.Error, LoggerEventIds.DisconnectHandlerError, "CreateDisconnectToken Callback");
+
+            private static readonly Action<ILogger, Exception?> _disconnectRegistrationCreateDisconnectTokenError =
+                LoggerMessage.Define(LogLevel.Error, LoggerEventIds.DisconnectRegistrationError, "CreateDisconnectToken");
+
+            private static readonly Action<ILogger, Exception?> _disconnectRegistrationError =
+                LoggerMessage.Define(LogLevel.Error, LoggerEventIds.DisconnectRegistrationError, "Unable to register for disconnect notifications.");
```
There were two log events that shared an event Id but with different messages. I wasn't sure whether to unify them or not, so I left them distinct for now.

davidfowl replied:
Oh you found a problem. We should unify them.

davidfowl replied:
Or use different ids

davidfowl commented on src/Servers/HttpSys/src/NativeInterop/RequestQueue.Log.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    internal partial class RequestQueue
+    {
+        private static class Log
```
For super small logs like this, its fine to put it in the same file

davidfowl commented on src/Servers/HttpSys/src/RequestProcessing/ClientCertLoader.Log.cs:
```diff
@@ -0,0 +1,30 @@
+        private static class Log
+        {
+            private static readonly Action<ILogger, Exception?> _channelBindingMissing =
+                LoggerMessage.Define(LogLevel.Error, LoggerEventIds.ChannelBindingMissing, "GetChannelBindingFromTls");
```
This message could use some work (I know its not related to this change)

martincostello replied:
Happy to update this PR with a more specific message for this if you have a suggestion.

davidfowl reviewed (approved):
Very nice. Few small nits.

martincostello committed:
Remove log partials with one message

Move the log partial classes that only contained one logger message directly into the classes that use them.

martincostello committed:
Split events that shared an event ID

Add new CreateDisconnectTokenError event ID to use for the less verbose variant of the log message that was also using DisconnectRegistrationError.

davidfowl commented on src/Servers/HttpSys/src/RequestProcessing/Request.cs:
```diff
@@ -20,6 +20,9 @@ namespace Microsoft.AspNetCore.Server.HttpSys
 {
     internal sealed class Request
     {
+        private static readonly Action<ILogger, Exception?> _errorInReadingCertificate =
```
Keep the static log class just make it a private class in this class.

davidfowl replied:
I wasn't clear on that, sorry...

martincostello replied:
No problem!

martincostello committed:
Restore nested Log classes

Restore 36de4d7d1fe09502ceb8e402b1f16005af6fb4fb but as nested classes in the same file, rather than a partial.

JamesNK commented on src/Servers/HttpSys/src/RequestProcessing/RequestContextOfT.Log.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace Microsoft.AspNetCore.Server.HttpSys
+{
+    internal sealed partial class RequestContext<TContext>
```
@davidfowl Realistically will there only ever be one context type used here at runtime?

I'm thinking about static properties on a generic type and property new instances being created for each generic parameter value.

davidfowl replied:
Only for value types. I don't think it'll matter here in practice but we can move it in another commit.

JamesNK replied:
Value types with generic parameters would create copies of IL, but the reference type delegates would still be copied between generic instances?

https://www.jetbrains.com/help/resharper/StaticMemberInGenericType.html

davidfowl replied:
Does this count as a member though? These are methods on a static nested class. I don't think that applies:

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQB4AVAPgYG8aGghgKH0mKBgGVsATwASMADaKIACgCUIwf2pC9DADIQA5gDpp8pStUAiACLYAbgEsAJjfUBuLQwC+PnwAHKGdHbAwYJnIkJlJDEx8dfVFGYmjxKVkFZTU0gAYGADtsfBhNXWS+H0qogE5VABIbbJU+YtLfD28K5P8evxo+mjFmADEICGFqJJSMuwgjY2MYKA1E6qEwqAYwBgBeIpgAd1Y2fK4NbpqthmB9w5P2CGAAKxgwDAuvDcEwcyyrGpvv09MB/pYcpcfH1fEA

davidfowl replied:
Oh wait, you're referring to the static delegates inside of the static class.

davidfowl replied:
```
System.NotSupportedException: Type C`1+Log[T] has a static constructor, which is not supported by SharpLab JIT decompiler.
 ```
Damn 😢 .

cc @benaadams do you happen to know?

Regardless, we can move it if it's a problem in theory, but there's a single T that hosting uses and it's currently a reference type. In general, I'd like to understand though 😄

JamesNK replied:
Sharplab doesn't like static ctors in generic types 🤕 

```cs
using System;

namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            var c = new C<string>();
            var b = new C<object>();
            c.SayHello();
            b.SayHello();
        }
    }

    public class C<T>
    {

        public void SayHello()
        {
            Log.SayHello("David");
        }

        private static class Log
        {
            public static readonly string Value = Guid.NewGuid().ToString();

            public static void SayHello(string name)
            {
                Console.WriteLine($"Hello {name} " + Value);
            }
        }
    }
}
```

![image](https://user-images.githubusercontent.com/303201/112772876-c123bd00-908f-11eb-8d60-c51382a7af7e.png)

davidfowl replied:
> Value types with generic parameters would create copies of IL, but the reference type delegates would still be copied between generic instances?

So `C<object>.Log._blah` and `C<string>.Log._blah` would end up being instantiated twice. Yes that makes sense and can be verified.

davidfowl replied:
@martincostello Can you send a follow up PR moving this specific log out of the generic type? It doesn't matter in practice but we can enforce this as a rule when looking at our logging codez.

JamesNK replied:
https://github.com/dotnet/aspnetcore/issues/31340

benaadams replied:
> cc @benaadams do you happen to know?

Here is much of muchness as are likely to only have one `TContext` per application and if you go via hosting then its always the same type. 

However yes the `static`s will be unique per generic regardless of whether they are reference or value types.

benaadams replied:
> However yes the `static`s will be unique per generic regardless of whether they are reference or value types.

*Aside: gives a really fast `IFeatureCollection.Get<T>()` style lookup; but alas it only works as a singleton lookup so can't use that for the actual `IFeatureCollection` since that's per request 😢

davidfowl replied:
> However yes the statics will be unique per generic regardless of whether they are reference or value types.

Yea, this is obvious in hindsight 😄

martincostello replied:
> @martincostello Can you send a follow up PR moving this specific log out of the generic type? It doesn't matter in practice but we can enforce this as a rule when looking at our logging codez.

Sure, I'll push something up today.

martincostello replied:
Someone beat me to it - #31345
